shared-libs/lineage/test/integration.js

```diff
@@ -0,0 +1,535 @@
+      }
+    }
+  },
+  reported_date: '5'
+};
+const report = {
+  _id: 'report',
+  type: 'data_record',
+  form: 'A',
+  contact: {
+    _id: report_contact._id,
+    parent: {
+      _id: report_parent._id,
+      parent: {
+        _id: report_grandparent._id
+      }
+    }
+  },
+  fields: {
+    patient_id: '12345'
+  }
+};
+const stub_contacts = {
+  _id: 'stub_contacts',
+  type: 'data_record',
+  contact: {
+    _id: 'something',
+    parent: {
+      _id: dummyDoc._id
+    }
+  },
+  form: {}
+};
+const stub_parents = {
+  _id: 'stub_parents',
+  type: 'clinic',
+  contact: { _id: 'something' },
+  parent: {
+    _id: 'something_else',
+    parent: { _id: 'dummyDoc' }
+  }
+};
+const sms_doc = {
+  _id: 'sms_doc',
+  type: 'data_record',
+  from: '123',
+  form: 'D',
+  fields: {},
+  sms_message: {
+    message_id: '76992',
+    message: 'somemessage',
+    from: '123',
+    type: 'sms_message',
+    form: 'D'
+  }
+};
+
+const fixtures = [
+  circular_area,
+  circular_chw,
+  circular_report,
+  dummyDoc,
+  dummyDoc2,
+  emptyObjectParent,
+  fetch_contact,
+  fetch_noContact,
+  no_contact,
+  no_lineageContact,
+  one_parent,
+  place_parent,
+  place_parentContact,
+  place_grandparent,
+  place_grandparentContact,
+  place_contact,
+  place,
+  report_parent,
+  report_parentContact,
+  report_grandparent,
+  report_grandparentContact,
+  report_contact,
+  report_patient,
+  report,
+  stub_contacts,
+  stub_parents,
+  sms_doc
+];
+const deleteDocs = ids => {
+  return db.allDocs({
+    keys: ids,
+    include_docs: true
+  }).then(data => {
+    const docs = data.rows.map(row => {
+      const doc = row.doc;
+      if (doc) {
+        doc._deleted = true;
+        return doc;
+      }
+    }).filter(doc => doc);
+    return db.bulkDocs(docs);
+  });
+};
+
+describe('Lineage', function() {
+  let lineage;
+
+  before(function() {
+    return memdownMedic('../../webapp/src')
+      .then(database => {
+        db = database;
+        lineage = lineageFactory(Promise, db);
+        return db.bulkDocs(fixtures);
+      });
+  });
+
+  after(function() {
+    const docIds = fixtures.map(doc => doc._id);
+    return deleteDocs(docIds);
+  });
+
+  afterEach(function() {
+    sinon.restore();
+  });
+
+  describe('fetchLineageById', function() {
+    it('returns correct lineage', function() {
+      return lineage.fetchLineageById(one_parent._id).then(result => {
+        chai.expect(result).to.have.lengthOf(2);
+        chai.expect(result[0]).excluding('_rev').to.deep.equal(one_parent);
+        chai.expect(result[1]).excluding('_rev').to.deep.equal(dummyDoc);
+      });
+    });
+  });
+
+  describe('fetchHydratedDoc', function() {
+    it('returns errors from query', function() {
+      return lineage.fetchHydratedDoc('abc').then(
+        () => {
+          throw new Error('should have thrown a 404');
+        },
+        err => {
+          chai.expect(err.status).to.equal(404);
+        }
+      );
```

Comment:
I haven't seen 2-arg `then()` for a while in our codebase; maybe more familiar to use `catch()`.

Reply:
Unfortunately necessary here, as if I do a `then()` followed by a `catch()`, the `catch` will catch the `Error('should have thrown a 404')` thrown by the `then`... unless you can think of a more sensible way to run the test!

Reply:
Neat!  I've definitely struggled with similar in the past 🙂